runtime/src/status_cache.rs

```diff
@@ -122,186 +162,361 @@ impl<T: Serialize + Clone> StatusCache<T> {
+                )
             });
 
-        // Update the max slot observed to contain txs using this blockhash.
-        *max_slot = std::cmp::max(slot, *max_slot);
+            // Update the max slot observed to contain txs using this blockhash.
+            max_slot.fetch_max(slot, Ordering::Relaxed);
 
-        // Grab the key slice.
-        let key_index = (*key_index).min(max_key_index);
-        let mut key_slice = [0u8; CACHED_KEY_SIZE];
-        key_slice.clone_from_slice(&key.as_ref()[key_index..key_index + CACHED_KEY_SIZE]);
+            // Grab the key slice.
+            let key_index = (*key_index).min(max_key_index);
+            unsafe {
+                ptr::copy_nonoverlapping(
+                    key.as_ref()[key_index..key_index + CACHED_KEY_SIZE].as_ptr(),
+                    key_slice.as_mut_ptr() as *mut u8,
+                    CACHED_KEY_SIZE,
+                )
+            }
+
+            // Insert the slot and tx result into the cache entry associated with
+            // this blockhash and keyslice.
+            let mut forks = txs.entry(unsafe { key_slice.assume_init() }).or_default();
+            forks.push((slot, res.clone()));
 
-        // Insert the slot and tx result into the cache entry associated with
-        // this blockhash and keyslice.
-        let forks = hash_map.entry(key_slice).or_default();
-        forks.push((slot, res.clone()));
+            key_index
+        };
 
-        self.add_to_slot_delta(transaction_blockhash, slot, key_index, key_slice, res);
+        self.add_to_slot_delta(
+            blockhash,
+            slot,
+            key_index,
+            unsafe { key_slice.assume_init() },
+            res,
+        );
     }
 
-    pub fn purge_roots(&mut self) {
+    fn purge_roots(&self) {
         if self.roots.len() > MAX_CACHE_ENTRIES {
-            if let Some(min) = self.roots.iter().min().cloned() {
+            if let Some(min) = self.roots().min() {
                 self.roots.remove(&min);
-                self.cache.retain(|_, (fork, _, _)| *fork > min);
+                self.cache.retain(|_key, value| {
+                    let (max_slot, _, _) = &**value;
+                    max_slot.load(Ordering::Relaxed) > min
+                });
                 self.slot_deltas.retain(|slot, _| *slot > min);
             }
         }
     }
 
-    /// Clear for testing
-    pub fn clear(&mut self) {
-        for v in self.cache.values_mut() {
-            v.2 = HashMap::new();
-        }
-
-        self.slot_deltas
-            .iter_mut()
-            .for_each(|(_, status)| status.lock().unwrap().clear());
+    #[cfg(feature = "dev-context-only-utils")]
+    pub fn clear(&self) {
+        self.cache.clear();
+        self.slot_deltas.clear();
     }
 
-    /// Get the statuses for all the root slots
+    /// Get the statuses for all the root slots.
+    ///
+    /// This is never called concurrently with add_root(), and for a slot to be a root there must be
```

Comment:
No concurrent calls with this right now. Not sure we can really prevent them from being added though.